File: practicas/practica_2/deepspace_Java_P2/src/main/java/deepspace/Weapon.java
/**
 * @file Weapon.java
 * @author Juan Manuel Rodríguez Gómez
 */
package deepspace;

/**
 * @brief Representa a las armas de las que puede disponer una estación espacial 
 *        para potenciar su energía al disparar
 */
class Weapon {

    private String name;
    private WeaponType type;
    private int uses;
    
    /**
     * @brief Constructor con parámetros
     * @param p_name Nombre
     * @param p_type Tipo de arma
     * @param p_uses Usos
     */
    Weapon(String p_name, WeaponType p_type, int p_uses) {
        name = p_name;
        type = p_type;
        uses = p_uses;
    }
    
    /**
     * @brief Constructor de copia
     * @param w Otro objeto de la clase Weapon
     */
    Weapon(Weapon w) {
        name = w.name;
        type = w.type;
        uses = w.uses;
    }
    
    /**
     * @brief Consultor del tipo de arma
     * @return type
     */
    public final WeaponType getType() {
        return type;
    }
    
    /**
     * @brief Consultor de los usos
     * @return uses
     */
    public final int getUses() {
        return uses;
    }

    /**
     * @brief Devuelve la potencia de disparo indicada por el tipo de arma
     */
    public final float power() {
        return type.getPower();
    }
    
    /**
     * @brief Emula el uso del arma
     */
    public final float useIt() {
        if(getUses() > 0) {
            uses--;
            return power();
        }
        else
            return 1.0f;
    }
    
    /**
     * @brief Construye una nueva instancia WeaponToUI a partir de la propia 
     *        instancia que recibe el mensaje y lo devuelve.
     *        Estos objetos constituyen una capa que permite conectar el 
     *        modelo con la interfaz de usuario (Weapon) manteniendo 
     *        cierto nivel de aislamiento entre ambos niveles
     * @return Instancia de la clase WeaponToUI
     */
    public WeaponToUI getUIversion() {
        return new WeaponToUI(this);
    }
    
    /**
     * @brief Función para representar una instancia de la clase Weapon
     *        en un string
     * @return s String que representa una instancia de la clase Weapon
     */
    public String toString() {
        
        String s;
        
        String s_name = "- Nombre del arma: " + name + " ";
        String s_type = "- Tipo de arma: " + type + " ";
        String s_power = "- Potencia de disparo del arma: " + power() + " ";
        String s_uses = "- Usos restantes del arma: " + uses + "\n";
        
        
        s = s_name + s_type + s_power + s_uses;
        
        return s;
    }
}

File: practicas/practica_2/Deepspace_Ruby_P2/lib/Weapon.rb
#encoding:utf-8

# File: Weapon.rb
# Author: Juan Manuel Rodríguez Gómez

module Deepspace

    # Brief: Representa a las armas de las que puede disponer una estación espacial 
    #        para potenciar su energía al disparar
    class Weapon

        # Brief: Constructor con parámetros
        # Param p_name: Nombre
        # Param p_type: Tipo de arma
        # Param p_uses: Usos
        def initialize(p_name, p_type, p_uses)
            @name = p_name
            @type = p_type
            @uses = p_uses
        end

        # Brief: Constructor de copia
        # Param w: Otro objeto de la clase Weapon
        def self.newCopy(w)
            copy = Weapon.new(w.name, w.type, w.uses)
        end

        public

        # Brief: Consultor del tipo de arma
        # Return: type
        def type
            if(@type == WeaponType::LASER)
                return WeaponType::LASER
            elsif (@type == WeaponType::MISSILE)
                return WeaponType::MISSILE
            else
                return WeaponType::PLASMA
            end
        end
      
        # Brief: Consultor del nombre del arma
        # Return: name
        def name
            @name
        end

        # Brief: Consultor de los usos
        # Return: uses
        def uses
            @uses
        end

        # Brief: Devuelve la potencia de disparo indicada por el tipo de arma
        def power
            return type.power
        end

        # Brief: Emula el uso del arma
        def useIt
            if (uses > 0)
              @uses -= 1
              return power
            else
              return 1.0
            end
        end

        # Brief: Construye una nueva instancia WeaponToUI a partir de la propia 
        #        instancia que recibe el mensaje y lo devuelve.
        #        Estos objetos constituyen una capa que permite conectar el 
        #        modelo con la interfaz de usuario (Weapon) manteniendo 
        #        cierto nivel de aislamiento entre ambos niveles
        # Return: Instancia de la clase WeaponToUI
        def getUIversion
            return WeaponToUI.new(self)
        end
        
        # Brief: Función para representar una instancia de la clase Weapon
        #        en un string
        # Return: String que representa una instancia de la clase Weapon
        def to_s
            s_name = "- Nombre del arma: #{@name} "
            s_type = "- Tipo de arma: #{@type.to_s} "
            s_power = "- Potencia de disparo del arma: #{power} "
            s_uses = "- Usos restantes del arma: #{@uses}\n"
            
            return s_name + s_type + s_power + s_uses
        end
    end # Class
end # Module